mtt/src/mtt.rs

```diff
@@ -0,0 +1,688 @@
+        let addr = phys_addr.bits();
+        // Bits 45:26
+        ((addr & 0x3fff_fc00_0000) >> 26) as usize
+    }
+
+    fn entry_for_addr(&mut self, phys_addr: SupervisorPageAddr) -> Result<MttL2Entry> {
+        use MttL2Entry::*;
+        let mtt_index = Self::get_mtt_index(phys_addr);
+        let value = self.mtt_l2_base[mtt_index];
+        if (value & L2_ZERO_MASK) != 0 {
+            return Err(InvalidL2Entry);
+        }
+        let mtt_l2_type = (value & L2_TYPE_MASK) >> L2_TYPE_SHIFT;
+
+        match mtt_l2_type {
+            // INFO must be 0 for 1GB entries
+            L2_1G_NC_TYPE if (value & L2_INFO_MASK) == 0 => Ok(NonConfidential1G(
+                Mtt1GEntry::new_nc(&mut self.mtt_l2_base[mtt_index..=mtt_index + 15]),
+            )),
+            L2_1G_C_TYPE if (value & L2_INFO_MASK) == 0 => Ok(Confidential1G(
+                Mtt1GEntry::new_conf(&mut self.mtt_l2_base[mtt_index..=mtt_index + 15]),
+            )),
+            L2_L1_TYPE => {
+                let pfn = Pfn::supervisor(value & L2_INFO_MASK);
+                // Safety: We are deferencing an existing L1 page entry in the MTT L2 table.
+                // This entry was created by Salus, or was created by trusted platform-FW,
+                // and the pfn corresponding to the page is guaranteed not be aliased and
+                // will remain valid for the lifetime of the L2 entry.
+                let l1_entry = unsafe { MttL1Entry::new(pfn) };
+                Ok(L1Entry(l1_entry))
+            }
+            L2_64M_TYPE => Ok(Mixed2M(Mtt64MEntry::new(
+                &mut self.mtt_l2_base[mtt_index..=mtt_index + 1],
+            ))),
+            _ => Err(InvalidL2Entry),
+        }
+    }
+}
+
+// Helper for iterating over a MTT range.
+struct MttRange<'a> {
+    addr: SupervisorPageAddr,
+    len: usize,
+    l2_dir: spin::MutexGuard<'a, MttL2Directory>,
+}
+
+type MttRangeCallback<'a> =
+    &'a mut dyn FnMut(SupervisorPageAddr, PageSize, MttL2Entry) -> ControlFlow<Result<()>, bool>;
+
+impl<'a> MttRange<'a> {
+    fn new(
+        addr: SupervisorPageAddr,
+        len: usize,
+        l2_dir: spin::MutexGuard<'a, MttL2Directory>,
+    ) -> Result<Self> {
+        const MAX_PAGE_ADDR: u64 = 0x3fff_ffff_f000;
+        if !PageSize::Size4k.is_aligned(len as u64) {
+            return Err(InvalidAddressRange);
+        }
+
+        let range_end = addr
+            .checked_add_pages(len as u64 / 4096)
+            .ok_or(InvalidAddressRange)?;
+        if range_end.bits() & !MAX_PAGE_ADDR != 0 {
+            return Err(InvalidAddressRange);
+        }
+
+        Ok(Self { addr, len, l2_dir })
+    }
+
+    // The callback will return ControlFlow::Continue with true to advance
+    // to the next address, or ControlFlow::Continue with false to retry
+    // with the same address. A ControlFlow::Break will terminate the loop.
+    fn range_iterator(&mut self, callback: MttRangeCallback) -> Result<()> {
```

Comment:
imo it might be cleaner to open code the loop rather than use the iterator at this point since we only need the retry logic for page splitting (which is what makes this bit complicated). but up to you.

i would suggest avoiding dynamic dispatch (i.e. use a generic rather than `dyn`) here, though.